sql-plugin/src/main/scala/com/nvidia/spark/rapids/GpuDeviceManager.scala

```diff
@@ -500,12 +442,110 @@ object GpuDeviceManager extends Logging {
+    val perTaskOverhead = conf.perTaskOverhead
+    val totalOverhead = perTaskOverhead * GpuDeviceManager.numCores
+    val confPinnedSize = conf.pinnedPoolSize
+    val confLimit = conf.offHeapLimit
+    // This min limit of 4GB is somewhat arbitrary, but based on some testing which showed
+    // that the previous minimum of 15 MB * num cores was too little for certain benchmark
+    // queries to complete, whereas this limit was sufficient.
+    val minMemoryLimit = 4L * 1024 * 1024 * 1024
+
+    val deviceCount: Int = Cuda.getDeviceCount()
+
+    val executorOverheadKey = "spark.executor.memoryOverhead"
+    val pysparkOverheadKey = "spark.executor.pyspark.memory"
+    val heapSizeKey = "spark.executor.memory"
+    val sparkOffHeapEnabledKey = "spark.memory.offHeap.enabled"
+    val sparkOffHeapSizeKey = "spark.memory.offHeap.size"
+
+    def toBytes = ConfHelper.byteFromString(_, ByteUnit.BYTE)
+
+    val executorOverhead = sparkConf.getOption(executorOverheadKey).map(toBytes)
+    val pysparkOverhead = toBytes(sparkConf.get(pysparkOverheadKey, "0"))
+    val heapSize = toBytes(sparkConf.get(heapSizeKey, "1g"))
+    val sparkOffHeapEnabled = sparkConf.getBoolean(sparkOffHeapEnabledKey, defaultValue = false)
+    val sparkOffHeapSize = if (sparkOffHeapEnabled) {
+      toBytes(sparkConf.get(sparkOffHeapSizeKey, "0"))
+    } else {
+      0L
+    }
+
+    val memoryLimit = if (confLimit.isDefined) {
+      if (executorOverhead.isEmpty) {
+        logWarning(s"$executorOverheadKey is not set")
+      }
+      logInfo(s"using configured ${RapidsConf.OFF_HEAP_LIMIT_SIZE} of ${confLimit.get}")
+      confLimit.get
+    } else {
+      // in case we cannot query the host for available memory due to environmental
+      // constraints, we can fall back to minMemoryLimit via saying there's no available
+      lazy val availableHostMemory = memCheck.getAvailableMemoryBytes(conf).getOrElse(0L)
+      val hostMemUsageFraction = .8
+      lazy val basedOnHostMemory = (hostMemUsageFraction * (availableHostMemory - heapSize
+        - pysparkOverhead - sparkOffHeapSize) / deviceCount).toLong
+      if (executorOverhead.isDefined) {
+        val basedOnConfiguredOverhead = executorOverhead.get - pysparkOverhead - sparkOffHeapSize
+        logWarning(s"${RapidsConf.OFF_HEAP_LIMIT_SIZE} is not set; we derived " +
+          s"a memory limit from ($executorOverheadKey - " + s"$pysparkOverheadKey -" +
+          s"$sparkOffHeapSizeKey) = (${executorOverhead.get} - $pysparkOverhead - " +
+          s"$sparkOffHeapSize) = $basedOnConfiguredOverhead")
+        if (basedOnConfiguredOverhead < minMemoryLimit) {
+          logWarning(s"memory limit $basedOnConfiguredOverhead is less than the minimum of " +
+            s"$minMemoryLimit; using the latter")
+          if (minMemoryLimit > basedOnHostMemory) {
+            logWarning(s"the amount of available memory detected on the host is " +
```

Comment:
I think we should include both `availableHostMemory` and `basedOnHostMemory` in this warning, in addition to `minMemoryLimit`. It is hard to word this one, but it would be nice to tell the user "you have $x available, and I computed $y as what I'd go to, but hey that's less than $z, so instead I am going to pick $z".